### PR TITLE
[SPARK-48848][PYTHON][DOCS] Set the upper bound version of `sphinxcontrib-*` in `dev/requirements.txt` with `sphinx==4.5.0`

### PR DESCRIPTION
**Pull request by wayneguow**
### What changes were proposed in this pull request?

This PR aims to set the upper bound version of 'sphinxcontrib-*' in `dev/requirements.txt` with `sphinx==4.5.0`.

### Why are the changes needed?

Currently, if Spark developers use the command `pip install --upgrade -r dev/requirements.txt` directly to install python-related dependencies, the automatically installed `sphinxcontrib-*` versions don't match `sphinx==4.5.0`. Refered to the issue: https://github.com/sphinx-doc/sphinx/issues/11890. 

Then they execute the `make html` command for building pySpark docs and the following error will appear:
image

This problem has been avoided through pinning `sphinxcontrib-*` in workflows of Spark GA:
image

So we can do the similar way by setting the upper bound version of in `requirements.txt` and it will be helpful for Spark developers when making pySpark docs.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Pass GA.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by wayneguow** [comment]
cc @HyukjinKwon

File: dev/requirements.txt [hunk 1]
```diff
@@ -37,6 +37,12 @@ nbsphinx
 numpydoc
 jinja2
 sphinx==4.5.0
+# With sphinx 4.5.0, we need to pin 'sphinxcontrib-*', it should unpin 'sphinxcontrib-*' after upgrading sphinx>5
+sphinxcontrib-applehelp == 1.0.4
```
Comment (HyukjinKwon) [inline dev/requirements.txt:41]: Can we set upperbound instead? This is for developement (not CI) so we don't really need to pin if it works.

File: dev/requirements.txt [hunk 2]
```diff
@@ -37,12 +37,13 @@ nbsphinx
...
-sphinxcontrib-htmlhelp == 2.0.1
-sphinxcontrib-qthelp == 1.0.3
-sphinxcontrib-serializinghtml == 1.1.5
+# With sphinx 4.5.0, we need to set the upperbound version of sphinxcontrib*, it should be removed after upgrading sphinx>=5
+sphinxcontrib-applehelp <= 1.0.4
+sphinxcontrib-devhelp <= 1.0.2
+sphinxcontrib-htmlhelp <= 2.0.1
+sphinxcontrib-jsmath <= 1.0.1
+sphinxcontrib-qthelp <= 1.0.3
+sphinxcontrib-serializinghtml <= 1.1.5
```
Comment (HyukjinKwon) [inline dev/requirements.txt:46]: can we remove spaces?

File: dev/requirements.txt (hunk 1 above)
Comment (wayneguow) [reply dev/requirements.txt]: I made a check that which versions are compatible with sphinx 4.x and set the upperbound version finally.  See the changelogs here:
https://github.com/sphinx-doc/sphinxcontrib-applehelp/blob/master/CHANGES
https://github.com/sphinx-doc/sphinxcontrib-devhelp/blob/master/CHANGES
https://github.com/sphinx-doc/sphinxcontrib-htmlhelp/blob/master/CHANGES
https://github.com/sphinx-doc/sphinxcontrib-jsmath/blob/master/CHANGES
https://github.com/sphinx-doc/sphinxcontrib-qthelp/blob/master/CHANGES
https://github.com/sphinx-doc/sphinxcontrib-serializinghtml/blob/master/CHANGES

File: dev/requirements.txt (hunk 2 above)
Comment (wayneguow) [reply dev/requirements.txt]: Fixed it.

File: dev/requirements.txt (hunk 1 above)
Comment (wayneguow) [reply dev/requirements.txt]: Issue title, PR title and description are also updated.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**